File: lib/set_metadata.dart
import 'dart:convert';
import 'dart:io';

import 'package:spot/track.dart' show Track;
import 'package:http/http.dart' as http;

void main() async {
  final content = File('tracks.json').readAsStringSync();
  final json = jsonDecode(content) as List<dynamic>;
  final tracks = json.map((dyn) => Track.fromJson(dyn));

  final images = Directory('images');

  if (!images.existsSync()) {
    images.createSync();
  }

  final metadata = Directory('metadata');

  if (!metadata.existsSync()) {
    metadata.createSync();
  }

  for (var track in tracks) {
    File? imageFile;

    try {
      imageFile = await downloadImage(track);
    } catch (e) {
      // noop
    }

    late File metaFile;

    try {
      metaFile = await saveMetadata(track);
    } catch (e) {
      print(e);
      continue;
    }

    late File trackFile;

    try {
      trackFile = await setMetadata(track, metaFile, imageFile);
    } catch (e) {
      print(e);
      continue;
    }

    try {
      await trackFile.rename(trackName(track));
      print(msg(track, 'OK!'));
    } on FileSystemException {
      print(msg(track, 'track rename error'));
    }
  }
}

Future<File> downloadImage(Track track) async {
  final file = File('images/${track.id}');

  if (await file.exists()) {
    return file;
  }

  final url = Uri.parse(track.albumImageUrl);
  final res = await http.get(url);

  await file.writeAsBytes(res.bodyBytes);

  return file;
}

Future<File> saveMetadata(Track track) async {
  List<String> rows = [];
  rows.add(buildTagRow('TITLE', track.name));
  rows.add(buildTagRow('ARTIST', track.artists));
  rows.add(buildTagRow('ALBUM', track.albumName));
  rows.add(buildTagRow('DATE', track.albumReleaseDate));
  rows.add(buildTagRow('COMMENT', track.spotifyUrl));

  final date = DateTime.tryParse('YYYY-MM-DD');

  if (date != null) {
    rows.add(buildTagRow('YEAR', date.year.toString()));
  }

  if (track.trackNumber != 0) {
    rows.add(buildTagRow('TRACKNUMBER', track.trackNumber.toString()));
  }

  if (track.totalTracks != 0) {
    rows.add(buildTagRow('TRACKNUMBER', track.totalTracks.toString()));
  }

  final file = File('metadata/${track.id}');
  await file.writeAsString('${rows.join('\n')}\n');

  return file;
}

String buildTagRow(String key, String value) {
  return '${key.toUpperCase()}=$value';
}

Future<File> setMetadata(Track track, File trackMetaFile, File? imageFile) async {
  final trackFile = File('${track.id}.flac');

  if (!await trackFile.exists()) {
    throw Exception(msg(track, 'track not found'));
  }

  List<String> parts = [];
  parts.add('--no-utf8-convert');
  parts.add('--remove-all-tags');

  if (await trackFile.exists()) {
    parts.add('--import-tags-from "${escaped(trackMetaFile.path)}"');
  } else {
    print(msg(track, 'metadata file not found'));
  }

  if (imageFile != null && await imageFile.exists()) {
    parts.add('--import-picture-from ${escaped(imageFile.path)}');
  } else {
    print(msg(track, 'image file not found'));
  }

  parts.add(escaped(trackFile.path));

  final cmdResult = await Process.run('metaflac', parts);

  if (cmdResult.exitCode == 0) {
    return trackFile;
  } else {
    throw Exception(msg(track, cmdResult.toString()));
  }
}

String escaped(String str) {
	return str.replaceAll('"', '\\"');
}

String trackName(Track track) {
  final artists = track.artists.replaceAll(';', ',');
  return '$artists - ${track.name}.flac';
}

String msg(Track track, String message) {
  return '${track.id} : $message';
}

File: src/set_metadata.rs
use crate::track::Track;
use anyhow::{Context, Result, bail};
use metaflac::{Tag, block::PictureType};
use reqwest::blocking;
use std::{
    fs,
    path::{Path, PathBuf},
};

pub fn call() -> Result<()> {
    let content = fs::read_to_string("tracks.json")?;
    let tracks: Vec<Track> = serde_json::from_str(&content)?;

    fs::create_dir_all("images")?;

    for track in tracks {
        if let Err(e) = process(&track) {
            eprintln!("{e}")
        }
    }

    Ok(())
}

fn process(track: &Track) -> Result<()> {
    let image_path = download_image(track).unwrap_or_default();
    let track_path = set_metadata(track, &image_path)?;
    let track_name = build_track_name(track);

    fs::rename(&track_path, &track_name).with_context(|| {
        let message = format!("track rename error: {}", &track_name);
        msg(track, &message)
    })?;
    println!("{}", msg(track, "OK!"));

    Ok(())
}

fn download_image(track: &Track) -> Result<PathBuf> {
    let path = Path::new("images").join(&track.id);

    if path.is_file() {
        return Ok(path);
    }

    let res = blocking::Client::new().get(&track.album_image_url).send()?;
    let image = res.bytes()?;

    fs::write(&path, image)?;

    Ok(path)
}

fn set_metadata(track: &Track, image_path: &Path) -> Result<PathBuf> {
    let p = format!("{}.flac", &track.id);
    let track_path = Path::new(&p);

    if !track_path.exists() {
        bail!(msg(track, "track not found"));
    }

    let mut tag = Tag::read_from_path(track_path).context(msg(track, "tag reading error"))?;

    // clear vorbis
    tag.remove_blocks(metaflac::BlockType::VorbisComment);

    set_vorbis_first(&mut tag, "TITLE", &track.name);
    set_vorbis_first(&mut tag, "ARTIST", &track.artists);
    set_vorbis_first(&mut tag, "ALBUM", &track.album_name);
    set_vorbis_first(&mut tag, "DATE", &track.album_release_date);
    set_vorbis_first(&mut tag, "COMMENT", &track.spotify_url);

    if let Some(year) = extract_year(&track.album_release_date) {
        set_vorbis_first(&mut tag, "YEAR", &year);
    }

    if track.track_number != 0 {
        let track_number = track.track_number.to_string();
        set_vorbis_first(&mut tag, "TRACKNUMBER", &track_number);
    }

    if track.total_tracks != 0 {
        let total_tracks = track.total_tracks.to_string();
        set_vorbis_first(&mut tag, "TOTALTRACKS", &total_tracks);
    }

    match fs::read(image_path) {
        Ok(picture) => match infer::get(&picture) {
            Some(mime_type) => {
                let picture_type = PictureType::CoverFront;
                tag.remove_picture_type(picture_type);
                tag.add_picture(mime_type.mime_type(), picture_type, picture);
            }
            None => {
                eprintln!("{}", msg(track, "unknown image format"));
            }
        },
        Err(_) => {
            eprintln!("{}", msg(track, "image file not found"));
        }
    }

    tag.save().context(msg(track, "tag saving error"))?;

    Ok(track_path.to_path_buf())
}

fn set_vorbis_first(tag: &mut Tag, key: &str, value: &str) {
    tag.set_vorbis(key, vec![value]);
}

// TODO: should sanitize symbols: \/:*?"<>| ?
fn build_track_name(track: &Track) -> String {
    let artists = track.artists.replace(";", ",");
    format!("{} - {}.flac", artists, track.name)
}

fn msg(track: &Track, message: &str) -> String {
    format!("{} : {}", track.id, message)
}

fn extract_year(date_str: &str) -> Option<String> {
    let parts: Vec<&str> = date_str.split('-').collect();

    if let Some(year) = parts.first()
        && year.chars().count() == 4
    {
        Some(year.to_string())
    } else {
        None
    }
}
